openfermioncirq/primitives/_ffft_test.py

```diff
@@ -0,0 +1,289 @@
+0   1     2   3     4   5     6   7
+│   │     │   │     │   │     │   │
+0↦0─1↦4───2↦1─3↦5───4↦2─5↦6───6↦3─7↦7
+│   │     │   │     │   │     │   │
+0↦0─1↦2───2↦1─3↦3   0↦0─1↦2───2↦1─3↦3
+│   │     │   │     │   │     │   │
+F₀──F₀    F₀──F₀    F₀──F₀    F₀──F₀
+│   │     │   │     │   │     │   │
+0↦0─1↦2───2↦1─3↦3   0↦0─1↦2───2↦1─3↦3
+│   │     │   │     │   │     │   │
+│   ω^0_4 │   ω^1_4 │   ω^0_4 │   ω^1_4
+│   │     │   │     │   │     │   │
+F₀──F₀    F₀──F₀    F₀──F₀    F₀──F₀
+│   │     │   │     │   │     │   │
+0↦0─1↦2───2↦1─3↦3   0↦0─1↦2───2↦1─3↦3
+│   │     │   │     │   │     │   │
+0↦0─1↦2───2↦4─3↦6───4↦1─5↦3───6↦5─7↦7
+│   │     │   │     │   │     │   │
+│   ω^0_8 │   ω^1_8 │   ω^2_8 │   ω^3_8
+│   │     │   │     │   │     │   │
+F₀──F₀    F₀──F₀    F₀──F₀    F₀──F₀
+│   │     │   │     │   │     │   │
+0↦0─1↦4───2↦1─3↦5───4↦2─5↦6───6↦3─7↦7
+│   │     │   │     │   │     │   │
+    """.strip()
+
+
+def test_ffft_fails_without_qubits():
+    with pytest.raises(ValueError):
+        ffft([])
+
+
+def test_ffft_fails_for_odd_size():
+    with pytest.raises(ValueError):
+        ffft(LineQubit.range(3))
+
+
+def test_inverse():
+    assert _inverse([0, 1, 2]) == [0, 1, 2]
+    assert _inverse([1, 2, 0]) == [2, 0, 1]
+    assert _inverse([2, 0, 1]) == [1, 2, 0]
+    assert _inverse([3, 2, 1, 0]) == [3, 2, 1, 0]
+
+
+@pytest.mark.parametrize(
+        'permutation, initial, expected',
+        [([0, 1], (1, []), (1, [])),
+         ([0, 1], (1, [1]), (1, [1])),
+         ([1, 0], (1, [1]), (1, [0])),
+         ([1, 2, 0], (1, [0]), (1, [1])),
+         ([1, 2, 0], (1, [1]), (1, [2])),
+         ([1, 2, 0], (1, [2]), (1, [0])),
+         ([1, 0], (1, [0, 1]), (-1, [0, 1])),
+         ([2, 1, 0], (1, [0, 2]), (-1, [0, 2])),
+         ([2, 1, 0], (1j, [0, 2]), (-1j, [0, 2])),
+         ([1, 0, 2], (1, [0, 2]), (1, [1, 2])),
+         ([0, 1, 2, 3, 4], (1, [0]), (1, [0])),
+         ([1, 0, 2, 3, 4], (1, [0]), (1, [1])),
+         ([2, 1, 0, 3, 4], (1, [0]), (1, [2])),
+         ([3, 1, 2, 0, 4], (1, [0]), (1, [3])),
+         ([4, 1, 2, 3, 0], (1, [0]), (1, [4])),
+         ([3, 1, 2, 0, 4], (1, [0, 4]), (1, [3, 4])),
+         ([3, 1, 2, 0, 4], (1, [0, 3]), (-1, [0, 3])),
+         ([3, 1, 2, 0, 4], (1, [0, 2]), (-1, [2, 3])),
+         ([0, 2, 4, 6, 1, 3, 5, 7], (1, [3, 4]), (-1, [1, 6])),
+         ([0, 2, 4, 6, 1, 3, 5, 7], (1, [2, 3, 4]), (1, [1, 4, 6])),
+         ([7, 6, 5, 4, 3, 2, 1, 0], (1, [2, 3, 4]), (-1, [3, 4, 5]))]
+)
+def test_permute(permutation, initial, expected):
+    n = len(permutation)
+    initial_state = _state_from_modes(n, *initial)
+    expected_state = _state_from_modes(n, *expected)
+    qubits = LineQubit.range(n)
+
+    ops = _permute(qubits, permutation)
+    circuit = cirq.Circuit.from_ops(ops)
+
+    state = circuit.apply_unitary_effect_to_state(
+        initial_state,
+        qubits_that_should_be_present=qubits
+    )
+
+    assert np.allclose(state, expected_state, rtol=0.0)
```

Comment:
Add a test that creates two equivalent circuits, one using `ffft` and one using `bogoliubov_transform`, and checks that the circuits are equivalent.

Reply:
Done.